File: frontend/src/app/services/analytics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { ApiResponse } from '../core/models/api-response.model';

// Adjust these import paths to match your actual model locations.
import { HeatmapDataDto } from '../models/analytics.model';
import { StatisticDto } from '../models/analytics.model';

// TODO: point this at your actual API base — swap for environment.apiUrl
const API_BASE = '/api/analytics';

/** Read-only analytics endpoints — all data is pre-computed by nightly aggregation jobs. */
@Injectable({ providedIn: 'root' })
export class AnalyticsService {
  constructor(private readonly http: HttpClient) {}

  /**
   * GET /api/analytics/top-danger-zones?periodEnd=...
   * Top 10 highest-severity accident hotspots. Defaults to yesterday's
   * data if periodEnd is omitted.
   */
  getTopDangerZones(periodEnd?: string): Observable<ApiResponse<HeatmapDataDto[]>> {
    const params = periodEnd ? { periodEnd } : undefined;
    return this.http.get<ApiResponse<HeatmapDataDto[]>>(`${API_BASE}/top-danger-zones`, { params });
  }

  /**
   * GET /api/analytics/statistics?periodType=&from=&to=
   * KPI snapshot for a period. Snapshots are generated at 00:30 nightly —
   * returns 404 if none exists yet for the requested period.
   */
  getStatistics(
    periodType: 'DAILY' | 'WEEKLY' | 'MONTHLY',
    from: string,
    to: string
  ): Observable<ApiResponse<StatisticDto>> {
    return this.http.get<ApiResponse<StatisticDto>>(`${API_BASE}/statistics`, {
      params: { periodType, from, to },
    });
  }

  /**
   * GET /api/analytics/heatmap?from=&to=
   * Pre-computed hotspot clusters for a date range; falls back to a live
   * query if the nightly job hasn't run yet. Feed straight into
   * Leaflet.js HeatLayer or Google Maps heatmap.
   */
  getHeatmap(from: string, to: string): Observable<ApiResponse<HeatmapDataDto[]>> {
    return this.http.get<ApiResponse<HeatmapDataDto[]>>(`${API_BASE}/heatmap`, { params: { from, to } });
  }
}

File: TrafficViolationSystem/src/main/java/com/academy/trafficviolationsystem/analytics/AnalyticsService.java
package com.academy.trafficviolationsystem.analytics;

import com.academy.trafficviolationsystem.core.exceptions.NotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Service for heatmap and statistics data.
 *
 * Report generation has been moved to ReportService — this service now
 * only handles heatmap and statistics queries so responsibilities are clean.
 *
 * Changes from original:
 *  - EntityNotFoundException (JPA) replaced with NotFoundException (core)
 *  - Report methods delegate to ReportService instead of duplicating logic
 *  - Principal replaced with UserPrincipal throughout
 */
@Service
@RequiredArgsConstructor
public class AnalyticsService {

    private final AnalyticsRepository            analyticsRepository;
    private final ViolationLocationLogRepository locationLogRepository;
    private final SystemStatisticsRepository     statisticsRepository;
    private final AnalyticsMapper                mapper;

    // ── Heatmap ───────────────────────────────────────────────────────────────

    /**
     * Returns pre-built hotspot data for the given date range.
     * Falls back to a live query from the raw location log if the nightly
     * ViolationAggregatorJob hasn't run yet for this period.
     */
    @Transactional(readOnly = true)
    public List<HeatmapDataDto> getHeatmapData(LocalDate from, LocalDate to) {
        List<AccidentHotspotEntity> hotspots =
                analyticsRepository.findByPeriodStartAndPeriodEnd(from, to);

        if (!hotspots.isEmpty()) {
            return hotspots.stream()
                    .map(mapper::toHeatmapDto)
                    .collect(Collectors.toList());
        }

        // Live fallback — build from raw log (nightly job hasn't run for this period)
        return locationLogRepository
                .clusterByGridCell(from.atStartOfDay(), to.plusDays(1).atStartOfDay(), 0.001)
                .stream()
                .map(row -> HeatmapDataDto.builder()
                        .latitude((Double) row[0])
                        .longitude((Double) row[1])
                        .violationCount(((Number) row[2]).intValue())
                        .severityScore(((Number) row[2]).doubleValue())
                        .dominantType((String) row[3])
                        .build())
                .collect(Collectors.toList());
    }

    // ── Statistics ────────────────────────────────────────────────────────────

    /**
     * Returns the pre-computed statistics snapshot for the given period.
     * Throws NotFoundException (not JPA EntityNotFoundException) if no
     * snapshot exists — ViolationAggregatorJob runs at 00:30 nightly.
     */
    @Transactional(readOnly = true)
    public StatisticsDto getStatistics(PeriodType periodType, LocalDate from, LocalDate to) {
        return statisticsRepository
                .findByPeriodTypeAndPeriodStartAndPeriodEnd(periodType, from, to)
                .map(mapper::toStatisticsDto)
                .orElseThrow(() -> new NotFoundException(
                        "No statistics snapshot found for " + periodType +
                                " " + from + " – " + to +
                                ". ViolationAggregatorJob generates snapshots nightly — try again after 00:30."));
    }

    // ── Top danger zones (used by dashboard) ──────────────────────────────────

    /**
     * Returns the 10 highest-severity hotspots for the most recent period end date.
     * Used by the dashboard to show the danger zone ranking card.
     */
    @Transactional(readOnly = true)
    public List<HeatmapDataDto> getTopDangerZones(LocalDate periodEnd) {
        return analyticsRepository
                .findTop10BySeverityScoreDescAndPeriodEnd(periodEnd)
                .stream()
                .map(mapper::toHeatmapDto)
                .collect(Collectors.toList());
    }
}
